### ISSUE TITLE
Bad typing after all function with None variable

### ISSUE THREAD

**Issue by same7m** [microsoft/pyright#2800]
**Describe the bug**
Values in an *all* block can't be inferred properly by pyright. 
```bash
test.py:5:11 - error: Operator "*" not supported for types "Literal[1]" and "None"
    Operator "*" not supported for types "Literal[1]" and "None" 
```
**To Reproduce**
Run pyright on the following code
```python
from typing import Optional
x: int = 1
y: Optional[int] = None
if all([x, y]):
    print(x*y)
```

**Expected behavior**
Shouldn't raise an error as after the *all* condition, variables can't be None.

**Comment by erictraut** [comment]
In this case, you are expecting `x` and `y` to be narrowed based on the expression `all([x, y])`. This isn't a supported type narrowing expression in pyright. For a full list of supported type narrowing forms, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards).

If you use one of the supported forms, the types will be narrowed as expected.

```python
x: int = 1
y: Optional[int] = None
if y is not None:
    print(x * y)
```

Mypy (another popular type checker) produces the same results as pyright in this case. 

We do occasionally add support for new type narrowing patterns, but each one requires special-case logic and adds complexity and performance overhead to analysis. We generally need to see some signal that it's sufficiently common to justify adding it. I haven't seen this particular form used before, and mypy doesn't support it. 

I'm going to close this issue for now, but we could revisit this in the future if enough people upvote (thumbs up) the issue.
